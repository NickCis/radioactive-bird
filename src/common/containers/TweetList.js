import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from 'material-ui/styles';
import connectWithSSR from '../components/connectWithSSR';
import { searchTweets } from '../actions/tweets';
import { CircularProgress } from 'material-ui/Progress';
import { withRouter } from 'react-router-dom';
import Tweet from '../components/Tweet';
import NotIdealState from '../components/NotIdealState';
import SentimentVeryDissatisfied from 'material-ui-icons/SentimentVeryDissatisfied';
import BugReport from 'material-ui-icons/BugReport';

const styles = {
  wrapper: {
    margin: '15px',
  },
  tweet: {
    margin: '0 auto',
  },
  loading: {
    'margin-top': '15px',
    width: '100%',
    'text-align': 'center',
  },
};

export class TweetList extends React.Component {
  static get propTypes() {
    return {
      searchTweets: PropTypes.func.isRequired,
      loading: PropTypes.bool.isRequired,
      tweets: PropTypes.array.isRequired,
      classes: PropTypes.object.isRequired,
      isError: PropTypes.bool,
      error: PropTypes.object,
    };
  }

  static getInitialData({ match, searchTweets }) {
    if (!match || !match.params || !match.params.query)
      return Promise.resolve();

    return searchTweets(match.params.query);
  }

  renderLoading() {
    const { classes } = this.props;
    return (
      <div className={classes.loading}>
        <CircularProgress />
      </div>
    );
  }

  renderEmpty() {
    return (
      <NotIdealState
        headline="Empty result"
        subheading="The search you have performed returned empty"
        Icon={SentimentVeryDissatisfied}
      />
    );
  }

  renderError() {
    const { error } = this.props;
    let errorText = 'An error was encountered while performing the search';

    if (error && error.errors && error.errors[0] && error.errors[0].message)
      errorText = error.errors[0].message;

    return (
      <NotIdealState headline="Error" subheading={errorText} Icon={BugReport} />
    );
  }

  render() {
    const { loading, isError, tweets, classes } = this.props;
    if (loading) return this.renderLoading();
    if (isError) return this.renderError();
    if (tweets.length === 0) return this.renderEmpty();

    return (
      <div className={classes.wrapper}>
        {tweets.map(t => (
          <Tweet key={t.id} className={classes.tweet} tweet={t} />
        ))}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  loading: state.search.loading,
  isError: state.search.isError,
  error: state.search.error,
  tweets: state.search.result.map(id => state.tweets[id]),
});

const mapDispatchToProps = dispatch => ({
  searchTweets: q => dispatch(searchTweets(q)),
});

export const StyledTweetList = withStyles(styles)(TweetList);
export default withRouter(
  connectWithSSR(mapStateToProps, mapDispatchToProps)(StyledTweetList)
);
