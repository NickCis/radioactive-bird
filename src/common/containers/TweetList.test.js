import React from 'react';
import { createRender } from 'material-ui/test-utils';
import { StyledTweetList as TweetList } from './TweetList';
import { StaticRouter } from 'react-router';

describe('<TweetList />', () => {
  const render = createRender();

  it('renders TweetList correctly - loading', () => {
    const wrapper = render(
      <StaticRouter location="/" context={{}}>
        <TweetList loading={true} searchTweets={jest.fn()} tweets={[]} />
      </StaticRouter>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders TweetList correctly - empty list', () => {
    const wrapper = render(
      <StaticRouter location="/" context={{}}>
        <TweetList loading={false} searchTweets={jest.fn()} tweets={[]} />
      </StaticRouter>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders TweetList correctly - with tweets', () => {
    const wrapper = render(
      <StaticRouter location="/" context={{}}>
        <TweetList
          loading={false}
          searchTweets={jest.fn()}
          tweets={[
            {
              id: 'test',
              title: 'test',
              user: {
                screen_name: 'test',
                name: 'Test',
              },
              favorite_count: 0,
              retweet_count: 0,
              text: 'test',
              created_at: 'Sun Sep 03 20:05:30 +0000 2017',
            },
          ]}
        />
      </StaticRouter>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
