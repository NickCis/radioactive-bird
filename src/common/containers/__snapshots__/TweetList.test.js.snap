// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TweetList /> renders TweetList correctly - empty list 1`] = `
<div
  class="TweetList-wrapper-1"
/>
`;

exports[`<TweetList /> renders TweetList correctly - loading 1`] = `
<div
  class="TweetList-loading-3"
>
  <div
    class="MuiCircularProgress-root-4 MuiCircularProgress-primaryColor-5"
    role="progressbar"
    style="width:40px;height:40px;"
  >
    <svg
      class="MuiCircularProgress-svg-7 MuiCircularProgress-indeterminateSvg-8"
      viewbox="0 0 40 40"
    >
      <circle
        class="MuiCircularProgress-circle-9 MuiCircularProgress-indeterminateCircle-10"
        cx="20"
        cy="20"
        fill="none"
        r="18.2"
        stroke-miterlimit="20"
        stroke-width="3.6"
      />
    </svg>
  </div>
</div>
`;

exports[`<TweetList /> renders TweetList correctly - with tweets 1`] = `
<div
  class="TweetList-wrapper-1"
>
  <div
    class="MuiPaper-root-19 MuiPaper-shadow2-23 MuiPaper-rounded-20 Tweet-card-12 TweetList-tweet-2"
  >
    <div
      class="MuiCardContent-root-51 MuiCardHeader-root-46"
    >
      <div
        class="MuiCardHeader-avatar-47"
      >
        <div
          class="MuiAvatar-root-52 MuiAvatar-colorDefault-53 Tweet-avatar-15"
        >
          t
        </div>
      </div>
      <div
        class="MuiCardHeader-content-48"
      >
        <span
          class="MuiTypography-root-55 MuiTypography-body2-63 MuiCardHeader-title-49"
        >
          Test @test
        </span>
        <span
          class="MuiTypography-root-55 MuiTypography-body2-63 MuiTypography-colorSecondary-75 MuiCardHeader-subheader-50"
        >
          September 3, 2017
        </span>
      </div>
    </div>
    <div
      class="MuiCardContent-root-51"
    >
      <p
        class="MuiTypography-root-55 MuiTypography-body1-64"
      >
        test
      </p>
    </div>
    <div
      class="MuiCardActions-root-77"
    >
      <button
        class="MuiButtonBase-root-88 MuiIconButton-root-79 Tweet-actionButton-17"
        role="button"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label-85"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-90"
            focusable="false"
            viewbox="0 0 24 24"
          >
            <path
              d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
            />
          </svg>
          <span
            class="MuiTypography-root-55 MuiTypography-caption-65"
          >
            0
          </span>
        </span>
        <span
          class="MuiTouchRipple-root-91"
        />
      </button>
      <button
        class="MuiButtonBase-root-88 MuiIconButton-root-79 Tweet-actionButton-17"
        role="button"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label-85"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-90"
            focusable="false"
            viewbox="0 0 24 24"
          >
            <path
              d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81 1.66 0 3-1.34 3-3s-1.34-3-3-3-3 1.34-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9c-1.66 0-3 1.34-3 3s1.34 3 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.16c-.05.21-.08.43-.08.65 0 1.61 1.31 2.92 2.92 2.92 1.61 0 2.92-1.31 2.92-2.92s-1.31-2.92-2.92-2.92z"
            />
          </svg>
          <span
            class="MuiTypography-root-55 MuiTypography-caption-65"
          >
            0
          </span>
        </span>
        <span
          class="MuiTouchRipple-root-91"
        />
      </button>
      <div
        class="Tweet-flexGrow-16"
      />
      <button
        aria-label="Go to Tweet"
        class="MuiButtonBase-root-88 MuiIconButton-root-79"
        role="button"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label-85"
        >
          <a
            class="Tweet-link-18"
            href="/tweet/test"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root-90"
              focusable="false"
              viewbox="0 0 24 24"
            >
              <path
                d="M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7c-2.76 0-5 2.24-5 5s2.24 5 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1zM8 13h8v-2H8v2zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4c2.76 0 5-2.24 5-5s-2.24-5-5-5z"
              />
            </svg>
          </a>
        </span>
        <span
          class="MuiTouchRipple-root-91"
        />
      </button>
    </div>
  </div>
</div>
`;
